pkg/server/server.go

```diff
@@ -83,40 +84,57 @@ func (s *srv) CreateUser(context.Context, *proto.CreateUserRequest) (*proto.Crea
 }
 
 func (s *srv) SetPassword(ctx context.Context, in *proto.SetPasswordRequest) (*proto.SetPasswordResponse, error) {
-	pw, err := s.serverAuthPassword()
-	if err != nil {
-		return nil, fmt.Errorf("getting manage auth password: %w", err)
-	}
-	if err := connection.CheckAuthFromContext(ctx, pw); err != nil {
-		return nil, fmt.Errorf("authorization failed: %w", err)
-	}
 	ds := datastore.New(s.config.datastoreReaderURL(), s.config.datastoreWriterURL())
 	auth := auth.New(s.config.authURL())
 	return setpassword.SetPassword(ctx, in, ds, auth)
 }
 
 func (s *srv) Tunnel(ts proto.Manage_TunnelServer) error {
-	pw, err := s.serverAuthPassword()
+	return tunnel.Tunnel(ts)
+}
+
+func serverAuthUnaryInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+	if err := serverAuth(ctx, info.Server.(*srv)); err != nil {
+		return nil, fmt.Errorf("server authentication: %w", err)
+	}
+	resp, err := handler(ctx, req)
 	if err != nil {
-		return fmt.Errorf("getting manage auth password: %w", err)
+		return nil, fmt.Errorf("calling handler: %w", err)
+	}
+	return resp, nil
+}
+
+func serverAuthStreamInterceptor(serv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+	if err := serverAuth(ss.Context(), serv.(*srv)); err != nil {
+		return fmt.Errorf("server authentication: %w", err)
 	}
-	if err := connection.CheckAuthFromContext(ts.Context(), pw); err != nil {
-		return fmt.Errorf("authorization failed: %w", err)
+	if err := handler(serv, ss); err != nil {
+		return fmt.Errorf("calling handler: %w", err)
 	}
-	return tunnel.Tunnel(ts)
+	return nil
 }
 
-func (s *srv) serverAuthPassword() ([]byte, error) {
-	dev, _ := strconv.ParseBool(s.config.OpenSlidesDevelopment)
-	if dev {
-		return []byte(shared.DevelopmentPassword), nil
+func serverAuth(ctx context.Context, s *srv) error {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return fmt.Errorf("getting metadata from context: failed")
 	}
-	p := s.config.PasswordFile
-	pw, err := os.ReadFile(p)
+	a := md.Get("authorization")
+	password, err := base64.StdEncoding.DecodeString(a[0])
 	if err != nil {
-		return nil, fmt.Errorf("reading manage auth password from secrets file %q: %w", p, err)
+		return fmt.Errorf("decoding password (base64): %w", err)
 	}
-	return pw, nil
+
+	sec, err := shared.ServerAuthSecret(s.config.PasswordFile, s.config.OpenSlidesDevelopment)
```

Comment:
Please cache the secret in the in `srv` object. You can do this in the `newServer` function or with [sync.Once](https://pkg.go.dev/sync#Once.Do) inside this function.